### PR TITLE
Fully working prox forest + tests

### PR DESCRIPTION
**Pull request by goastler**
Redo of the prox-forest branch which went very off the rails with merging recent workshop changes.

Hence this branch is a cut / paste job to fix the merge issues.

Changes are as follows:

- altered parameter space for proximity forest (should fix accuracy issues)
- added TWE in cython
- optimised cython
- proximity forest testing

Todo:

- docstrings need to be updated (they're currently the old version for PF v1)
- proper unit testing rather than functionality testing

### CODE REVIEW

File: sktime/utils/dataset_properties.py [hunk 1]
```diff
@@ -30,12 +47,25 @@ def bin_instances_by_class(instances, class_labels):
         instances_bin.append(instance)
     return bins
 
-# find the maximum length of an instance from a set of instances
-def max_instance_length(instances, dimension = 0):
+
+# find the maximum length of an instance from a set of instances for a given dimension
+def max_instance_dimension_length(instances, dimension):
     num_instances = instances.shape[0]
     max = -1
```
Comment (mloning) [inline sktime/utils/dataset_properties.py]: would rename `max` here and below to avoid overwriting Python function

File: sktime/utils/dataset_properties.py [hunk 2]
```diff
@@ -1,16 +1,33 @@
...
+
+
+# set dataframe indices to a positive range (0 to above 0)
+def positive_dataframe_indices(X):
+    if np.any(X.index < 0) or len(np.unique(X.index)) > 1:
+        X = X.copy(deep=True)
+        X.index = np.arange(0, len(X.index))
+    return X
+
 # find the standard deviation of the dataset
```
Comment (mloning) [inline sktime/utils/dataset_properties.py]: would rename `instances` here and throughout to `X` for consistency

File: sktime/classifiers/tests/test_proximity.py [hunk 3]
```diff
@@ -13,520 +14,508 @@ def run_classifier_on_dataset(classifier, dataset_loader,
...
-
-def test_proximity_stump_on_gunpoint():
-    classifier = ProximityStump(debug = True, random_state = 0)
-    run_classifier_on_dataset(classifier, load_gunpoint, stump_gunpoint_predict_probas, stump_gunpoint_predictions)
-
+def test_proximity_forest_on_gunpoint():
+    classifier = ProximityForest(random_state=0, n_trees=3, find_stump=proximity.best_of_n_stumps(1))
+    run_classifier_on_dataset(classifier, load_gunpoint, forest_gunpoint_predict_probas, forest_gunpoint_predictions)
 
 stump_gunpoint_predict_probas = np.array([
```
Comment (mloning) [inline sktime/classifiers/tests/test_proximity.py]: perhaps only run a few samples of gunpoint?

File: sktime/transformers/series_to_series.py [hunk 4]
```diff
@@ -433,3 +430,35 @@ def transform(self, X, y=None):
         Xt["%s_starts" % column_prefix] = pd.Series(self._starts)
         Xt["%s_lengths" % column_prefix] = pd.Series(self._lengths)
         return Xt
+
+
+class CachedTransformer(BaseTransformer):
```
[the remark this answers is not in the document]
Comment (goastler) [reply sktime/transformers/series_to_series.py]: I'll add the docstrings :)

All it does is cache individual transformations of single instances. For instance, PF needs the derivative for certain distance measures. By nature, PF won't necessarily need all the data transformed, especially if it's a large dataset. Hence this transformer does the transformations on demand :)

File: sktime/classifiers/tests/test_proximity.py (hunk 3 above)
Comment (goastler) [reply sktime/classifiers/tests/test_proximity.py]: The tests made quite a problem for me. Currently, they test PF / PT / PS well, but take a while. If I trim the dataset, I risk making the problem too simple and not thoroughly testing the classifiers, though the tests would be quicker.

I don't know which way is both, what's your thoughts @mloning @jasonlines @TonyBagnall ?

File: sktime/transformers/series_to_series.py [hunk 5]
```diff
@@ -1,15 +1,12 @@
...
+from sklearn.utils.validation import check_random_state
 
-from sktime.utils.validation import check_equal_index, check_ts_array
-from sktime.utils.transformations import tabularize, detabularize, concat_nested_arrays
 from sktime.transformers.base import BaseTransformer
+from sktime.utils.transformations import tabularize, detabularize, concat_nested_arrays
+from sktime.utils.validation import check_equal_index
 
-
-__all__ = ['RandomIntervalSegmenter', 'IntervalSegmenter', 'DerivativeSlopeTransformer', 'TimeSeriesConcatenator',
```
[the remark this answers is not in the document]
Comment (goastler) [reply sktime/transformers/series_to_series.py]: sorry I think my IDE did that

File: sktime/classifiers/tests/test_proximity.py (hunk 3 above)
Comment (mloning) [reply sktime/classifiers/tests/test_proximity.py]: I prefer quicker for unit tests, we do test them for equivalence against the java code base etc

File: sktime/transformers/series_to_series.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mloning) [reply sktime/transformers/series_to_series.py]: Okay, not sure if I can follow, is this specific to prox forest, then I would put it in that file for now ...

File: sktime/classifiers/tests/test_proximity.py (hunk 3 above)
Comment (TonyBagnall) [reply sktime/classifiers/tests/test_proximity.py]: yes trim, the tests are not for the quality of the algorithm, just that it builds ok
